### PR TITLE
  Add SentinelKilnDB dataset

### LINKED ISSUE #3421: New dataset: sentinelkiln-dataset
### Summary

- Source: [kaggle](https://www.kaggle.com/datasets/rishabhsnip/sentinelkiln-dataset)
- License: https://creativecommons.org/licenses/by-nc/4.0/
- Format: OBB - this is new to torchgeo?

### Rationale

_No response_

### Implementation

_No response_

### Alternatives

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by rishabh-mondal**
Description:
  ## Summary

1.  [SentinelKilnDB: A Large-Scale Dataset and Benchmark for OBB Brick Kiln Detection in South Asia Using Satellite Imagery](https://neurips.cc/virtual/2025/loc/san-diego/poster/121530)
2.  Adds SentinelKilnDB dataset for brick kiln detection using Sentinel-2 imagery
3. 114K+ images (128x128 RGB) with 3 kiln classes (CFCBK, FCBK, Zigzag)
4. Supports both axis-aligned and oriented bounding boxes
5. Paper accepted at NeurIPS 2025

  ## References
  1. Dataset: https://huggingface.co/datasets/SustainabilityLabIITGN/SentinelKilnDB
  2. Paper: https://openreview.net/forum?id=efGzsxVSEC
  3. https://neurips.cc/virtual/2025/loc/san-diego/poster/121530
  
  ## Test plan
  - [x] Unit tests pass (`pytest tests/datasets/test_sentinelkilndb.py -v` - 33 tests)
  - [x] Linting passes (`ruff check torchgeo/datasets/sentinelkilndb.py`)
  - [x] Type checking passes (`ty check torchgeo/datasets/sentinelkilndb.py`)
  - [x] Verified with real data download (train: 65K, validation: 24K, test: 18K samples)

## Summary by CodeRabbit

* **New Features**
  * Added the SentinelKilnDB dataset with automatic download, train/val/test splits, and support for horizontal and oriented bounding boxes; dataset is now exported for public use.
  * Built-in visualization utilities for inspecting samples and boxes.

* **Tests**
  * Added comprehensive test suite covering splits, orientations, error cases, plotting, and download/checksum behaviors.
  * Included a deterministic test data generator to produce representative and edge-case dataset samples.

* **Documentation**
  * Added API docs entry for SentinelKilnDB.

Closes #3421

### CODE REVIEW

File: tests/data/sentinelkilndb/data.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+import io
+import os
+from pathlib import Path
+
+import numpy as np
+import pandas as pd
+from PIL import Image
+
+
+def create_dummy_image_bytes(size: tuple[int, int] = (128, 128)) -> bytes:
```
Comment (adamjstewart) [inline tests/data/sentinelkilndb/data.py]: Can we make the fake test files smaller? Maybe 32x32 to save space in the repo.

File: tests/data/sentinelkilndb/data.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        # Create DataFrame and save as parquet
+        df = pd.DataFrame(rows)
+        parquet_path = root / f'{split}.parquet'
+        df.to_parquet(parquet_path)
+        print(f'Created {parquet_path}')
+
+        # Calculate MD5
+        with open(parquet_path, 'rb') as f:
+            md5 = hashlib.md5(f.read()).hexdigest()
+        print(f'  MD5: {md5}')
```
Comment (adamjstewart) [inline tests/data/sentinelkilndb/data.py]: You can remove this, we no longer test checksums for fake data.

File: torchgeo/datasets/sentinelkilndb.py [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+import pandas as pd
+import torch
+from matplotlib import patches
+from matplotlib.figure import Figure
+from PIL import Image
+from torch import Tensor
+
+from .errors import DatasetNotFoundError
+from .geo import NonGeoDataset
+from .utils import Path, Sample, download_url, percentile_normalization
```
Comment (adamjstewart) [inline torchgeo/datasets/sentinelkilndb.py]: percentile_normalization is now deprecated, can you switch to quantile_normalization?

File: torchgeo/datasets/sentinelkilndb.py [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+class _FileInfo(TypedDict):
+    filename: str
+    md5: str | None
+
+
+class SentinelKilnDB(NonGeoDataset):
+    """SentinelKilnDB brick kiln detection dataset.
+
+    *SentinelKilnDB: A Large-Scale Dataset and Benchmark for OBB Brick Kiln Detection in South Asia Using Satellite Imagery*
+    Accepted at NeurIPS 2025.
```
Comment (adamjstewart) [inline torchgeo/datasets/sentinelkilndb.py]: Newlines get stripped, so this will end up as one long sentence if you don't add punctuation

File: torchgeo/datasets/sentinelkilndb.py [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+
+    * https://openreview.net/forum?id=efGzsxVSEC
+
+    .. note::
+
+       This dataset requires the following additional library to be installed:
+
+       * `pyarrow <https://pypi.org/project/pyarrow/>`_: to load parquet files
+
+    .. versionadded:: 0.8
```
Comment (adamjstewart) [inline torchgeo/datasets/sentinelkilndb.py]:
```suggestion
    .. versionadded:: 0.10
```

File: torchgeo/datasets/sentinelkilndb.py [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+    """
+
+    url = 'https://huggingface.co/datasets/SustainabilityLabIITGN/SentinelKilnDB/resolve/refs%2Fconvert%2Fparquet/default/{}/0000.parquet'
```
Comment (adamjstewart) [inline torchgeo/datasets/sentinelkilndb.py]: Any reason to use the `refs/convert/parquet` branch instead of the main branch? We'll need to change this to a specific git commit hash for stability.

File: torchgeo/datasets/sentinelkilndb.py [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+
+    file_info: ClassVar[dict[str, _FileInfo]] = {
+        'train': {'filename': 'train.parquet', 'md5': None},
+        'validation': {'filename': 'validation.parquet', 'md5': None},
+        'test': {'filename': 'test.parquet', 'md5': None},
+    }
```
Comment (adamjstewart) [inline torchgeo/datasets/sentinelkilndb.py:83]: Is this attribute useful? The filename can always be easily guessed from the split name.

File: torchgeo/datasets/sentinelkilndb.py [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+    valid_splits = ('train', 'validation', 'test')
+    valid_orientations = ('horizontal', 'oriented')
+
+    # Fixed image size for this dataset
+    image_size = 128
+
+    def __init__(
+        self,
+        root: Path = 'data',
+        split: Literal['train', 'validation', 'test'] = 'train',
```
Comment (adamjstewart) [inline torchgeo/datasets/sentinelkilndb.py]: We usually use train/val/test as split names for consistency.

File: torchgeo/datasets/sentinelkilndb.py (hunk 7 above)
Comment (adamjstewart) [reply torchgeo/datasets/sentinelkilndb.py:83]: Why is the md5 always None? We should use the actual MD5s here.

File: torchgeo/datasets/sentinelkilndb.py [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+        os.makedirs(self.root, exist_ok=True)
+
+        filename = self.file_info[self.split]['filename']
+        md5 = self.file_info[self.split]['md5'] if self.checksum else None
+
+        download_url(
+            url=self.url.format(self.split), root=self.root, filename=filename, md5=md5
+        )
+
+    def plot(
```
Comment (adamjstewart) [inline torchgeo/datasets/sentinelkilndb.py:319]: Can you share some sample plots of real data (horizontal, oriented, positive, negative) so we can see what they look like?

File: torchgeo/datasets/sentinelkilndb.py (hunk 3 above)
Comment (adamjstewart) [inline torchgeo/datasets/sentinelkilndb.py]: percentile_normalization is now deprecated, can you switch to quantile_normalization?

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Needs docs:

- [ ] docs/api/datasets/sentinelkilndb.rst
- [ ] docs/api/datasets/benchmark.csv

A bit confused why the data is stored as strings in parquet instead of floats...

**Comment by rishabh-mondal** [comment]
## Addressed Review Comments

Thanks for the detailed review! I've addressed all the comments:

| Comment | Status |
|---------|--------|
| Use 32x32 test images | ✅ Done |
| Remove MD5 calculation from test data | ✅ Removed |
| Switch to `quantile_normalization` | ✅ Done |
| Add punctuation to docstring title | ✅ Added period |
| Update versionadded to 0.10 | ✅ Fixed |
| Use main branch with commit hash | ✅ Using main branch with SHA256 checksums |
| Use train/val/test split names | ✅ Changed `validation` to `val` |
| Add actual checksums | ✅ Added SHA256 from HuggingFace LFS |

### Sample Plots

Here are sample visualizations showing FCBK, Zigzag, CFCBK, Mixed Classes, and Negative samples with both horizontal and oriented bounding boxes:

SentinelKilnDB Sample Visualizations

**Comment by calebrob6** [comment]
Looks like a small ty issue

**Comment by rishabh-mondal** [comment]
I solved the issues. Could you please let me know any further changes needed or not ?

**Comment by calebrob6** [comment]
Just need to add a test that hits the `checksum=True` case, e.g here's one that I used in DLRSD in  https://github.com/torchgeo/torchgeo/pull/3498.
```
    def test_corrupted(self, tmp_path: Path) -> None:
        with open(os.path.join(tmp_path, 'DLRSD.zip'), 'w') as f:
            f.write('bad')
        with pytest.raises(RuntimeError, match='Dataset found, but corrupted'):
            DLRSD(root=tmp_path, checksum=True)
```

File: docs/api/datasets/benchmark.csv [hunk 10]
```diff
@@ -55,6 +55,7 @@ Dataset,Task,Source,# Samples,# Classes,Size (px),Resolution (m),Bands,License
 :ref:`SSL4EO-L Benchmark`,S,Lansat & CDL,25K,134,264x264,30,MSI,CC0-1.0
 :ref:`SSL4EO-L Benchmark`,S,Lansat & NLCD,25K,17,264x264,30,MSI,CC0-1.0
 :ref:`SeasoNet`,S,Sentinel-2,"1,759,830",33,120x120,10,MSI,CC-BY-4.0
+:ref:`SentinelKilnDB`,OD,Sentinel-2,"114,400",3,128x128,10,RGB,CC-BY-NC-4.0
```
Comment (adamjstewart) [inline docs/api/datasets/benchmark.csv:60]: Confirmed license from https://huggingface.co/datasets/SustainabilityLabIITGN/SentinelKilnDB

File: torchgeo/datasets/sentinelkilndb.py [hunk 11]
```diff
@@ -0,0 +1,420 @@
...
+    .. note::
+
+       This dataset requires the following additional library to be installed:
+
+       * `pyarrow <https://pypi.org/project/pyarrow/>`_: to load parquet files
+
+    .. versionadded:: 0.10
+    """
+
+    url = 'https://huggingface.co/datasets/SustainabilityLabIITGN/SentinelKilnDB/resolve/main/{0}/{0}.parquet'
```
Comment (adamjstewart) [inline torchgeo/datasets/sentinelkilndb.py]: Needs a stable git commit hash to ensure stable downloads:

```suggestion
    url = 'https://hf.co/datasets/SustainabilityLabIITGN/SentinelKilnDB/resolve/48242579f58fbec5b40316f9eb03aac573676f2e/{0}/{0}.parquet'
```

File: torchgeo/datasets/sentinelkilndb.py [hunk 12]
```diff
@@ -0,0 +1,420 @@
...
+            tuple of boxes (N, 4) and labels (N,)
+        """
+        boxes = []
+        labels = []
+
+        if len(label_array) == 0:
+            return (
+                torch.zeros((0, 4), dtype=torch.float32),
+                torch.zeros(0, dtype=torch.long),
+            )
```
Comment (adamjstewart) [inline torchgeo/datasets/sentinelkilndb.py:225]: Are we returning an empty bounding box for negative samples? I don't think we should return any bounding box in that case.

File: torchgeo/datasets/sentinelkilndb.py [hunk 13]
```diff
@@ -0,0 +1,420 @@
...
+            download_url(
+                url=self.url.format(self.split),
+                root=self.root,
+                filename=filename,
+                sha256=sha256,
+            )
+        else:
+            download_url(
+                url=self.url.format(self.split), root=self.root, filename=filename
+            )
```
Comment (adamjstewart) [inline torchgeo/datasets/sentinelkilndb.py]: Could also use inline branching and a single call:
```python
download_url(
    url=...
    root=...
    filename=...
    sha256=sha256 if self.checksum else None,
)
```
Saves some lines of code.

File: sentinelkilndb_samples.png
Comment (adamjstewart) [inline sentinelkilndb_samples.png:1]: I think this file was accidentally added to the PR, this can be removed.

File: torchgeo/datasets/sentinelkilndb.py [hunk 14]
```diff
@@ -0,0 +1,420 @@
...
+        if len(label_array) == 0:
+            return (
+                torch.zeros((0, 4), dtype=torch.float32),
+                torch.zeros(0, dtype=torch.long),
+            )
+
+        for line in label_array:
+            parts = line.strip().split()
+            if len(parts) < 5:
+                continue
```
Comment (adamjstewart) [inline torchgeo/datasets/sentinelkilndb.py]: Does this ever happen in the real data, or is this defensive coding for a situation that will never arise?

**Review by adamjstewart — CHANGES_REQUESTED** [review]
A few minor comments, otherwise this looks good to go.

File: torchgeo/datasets/sentinelkilndb.py [hunk 15]
```diff
@@ -0,0 +1,420 @@
...
+                url=self.url.format(self.split), root=self.root, filename=filename
+            )
+
+    def plot(
+        self,
+        sample: Sample,
+        show_titles: bool = True,
+        suptitle: str | None = None,
+        box_alpha: float = 0.7,
+    ) -> Figure:
```
[the remark this answers is not in the document]
Comment (adamjstewart) [reply torchgeo/datasets/sentinelkilndb.py:325]: This looks correct, we should add titles to plots or something.

**Comment by rishabh-mondal** [comment]
**Re: Empty bbox for negative samples**

For negative samples, we return `torch.zeros((0, 4))` for boxes and `torch.zeros(0, dtype=torch.long)` for labels. This is the standard object detection convention - shape `(0, 4)` indicates zero boxes while preserving tensor structure for batching.

```python
# Positive: 2 boxes
boxes = torch.tensor([[10, 20, 50, 80], [90, 40, 140, 120]])
print(boxes.shape)  # torch.Size([2, 4])

# Negative: 0 boxes
boxes = torch.zeros((0, 4))
print(boxes.shape)  # torch.Size([0, 4])
```

**Comment by rishabh-mondal** [comment]
**Re: Line 230 - Defensive coding question**

This is defensive coding. The dataset labels are well-formed, but this guard ensures robustness against any edge cases.

**Comment by rishabh-mondal** [comment]
**Re: Inline branching for download_url**

Done. Simplified to use inline branching:

```python
sha256 = self.file_info[self.split]['sha256'] if self.checksum else None

download_url(
    url=self.url.format(self.split),
    root=self.root,
    filename=filename,
    sha256=sha256,
)
```

**Comment by rishabh-mondal** [comment]
**Re: Add titles to plots**

Done. The `show_titles` flag now controls whether box labels and suptitle are displayed. When `show_titles=False`, no text annotations are added to the plot.

**Comment by rishabh-mondal** [comment]
**Re: Remove accidentally added file**

Done. Removed `sentinelkilndb_samples.png` from the PR.

**Comment by rishabh-mondal** [comment]
**Re: Use stable git commit hash**

Done. Updated URL to use commit hash `a467e64329b486fadf236cde2db8a2bbf85fec6c` for stable downloads.

**Comment by adamjstewart** [comment]
> **Re: Line 230 - Defensive coding question**
> 
> This is defensive coding. The dataset labels are well-formed, but this guard ensures robustness against any edge cases.

The dataset labels aren't going to change, no need for handling of edge cases that don't exist.

**Comment by rishabh-mondal** [comment]
**Re: Defensive coding**

Done. Removed the defensive checks (`if len(parts) < 5/9: continue`) from both `_parse_yolo_aa` and `_parse_yolo_obb` methods since the dataset labels are well-formed.

File: torchgeo/datasets/sentinelkilndb.py (hunk 12 above)
Comment (adamjstewart) [reply torchgeo/datasets/sentinelkilndb.py:225]: Update: #3534 suggests this may actually be a good idea, waiting for the discussion there to be finalized.

**Comment by adamjstewart** [comment]
Just needs additional test coverage now. Could also wait until we finalize the discussion in #3534 for how to handle images with no boxes.

**Comment by rishabh-mondal** [comment]
Thanks, @adamjstewart. I am also very interested in contributing to other TorchGeo issues. Please point me to some issues I could work on.

**Comment by robmarkcole** [comment]
@rishabh-mondal 3 lines of code require test coverage, going to pick this up?

**Comment by calebrob6** [comment]
+1 to the test coverage, this seems very close to merging @rishabh-mondal
